### PR TITLE
Make static Store access shard lock aware

### PR DESCRIPTION
**Pull request by bleskes**
We currently have concurrency issue between the static methods on the Store class and store changes that are done via a valid open store. An example of this is the async shard fetch which can reach out to a node while a local shard copy is shutting down (the fetch does check if we have an open shard and tries to use that first, but if the shard is shutting down, it will not be available from IndexService).

Specifically, async shard fetching tries to read metadata from store, concurrently the shard that shuts down commits to lucene, changing the segments_N file. this causes a file not find exception on the shard fetching side. That one in turns makes the master think the shard is unusable. In tests this can cause the shard assignment to be delayed (up to 1m) which fails tests. See https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+java9-periodic/570 for details.

This is one of the things #18938 caused to bubble up.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/indices/store/TransportNodesListShardStoreMetaData.java [hunk 1]
```diff
@@ -150,7 +150,10 @@ private StoreFilesMetaData listStoreMetaData(ShardId shardId) throws IOException
             if (shardPath == null) {
                 return new StoreFilesMetaData(shardId, Store.MetadataSnapshot.EMPTY);
             }
-            return new StoreFilesMetaData(shardId, Store.readMetadataSnapshot(shardPath.resolveIndex(), shardId, logger));
+            // note that this may fail if it can't get access to the shard lock. Since we check above there is an active shard, this means:
+            // 1) a shard is being constructed, which means the master will not use a copy of this replcia
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/indices/store/TransportNodesListShardStoreMetaData.java]: Did you mean to add a `2)` here?

Comment (bleskes) [inline core/src/main/java/org/elasticsearch/indices/store/TransportNodesListShardStoreMetaData.java]: hehe, yeah - ADD or the works :)
2) A shard is shutting down and has not cleared it's content within lock timeout. In this case the master may not reuse local resources.

File: core/src/main/java/org/elasticsearch/index/store/Store.java [hunk 2]
```diff
@@ -1398,4 +1402,13 @@ private static long estimateSize(Directory directory) throws IOException {
         }
     }
 
+    /**
+     * A shard lock supplier that is used by the static methods on this class. Normal methods rely on
+     * the shard lock passed to the constructor.
+     */
+    @FunctionalInterface
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/store/Store.java]: can we move this into NodeEnv instead?

**Comment by s1monw** [comment]
left on  nit - LGTM otherwise

**Comment by ywelsch** [comment]
LGTM

File: core/src/main/java/org/elasticsearch/index/store/Store.java (hunk 2 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/index/store/Store.java]: sure

**Pull request merged**